File: scripts/javascript/05_coral_reefs_aca.js
/**********************************************************************
 * 05_coral_reefs_aca.js   —  CORAL REEF HABITATS (Allen Coral Atlas)
 * SPREP / UNEP GEE Climate Workshop 2026
 *
 * GOAL: Map the coral reefs and seafloor habitats around YOUR islands —
 *       where the coral, seagrass, sand and reef structures are — and
 *       measure how much coral habitat your country has.
 *
 * DATASET: Allen Coral Atlas (ACA/reef_habitat/v2_0)
 *   A single ee.Image at 5 m, mapped from satellite imagery 2018-2021.
 *   Bands used here:
 *     'benthic'    seafloor cover: Sand, Rubble, Rock, Seagrass,
 *                  Coral/Algae, Microalgal Mats
 *     'geomorphic' reef zone: reef flat, crest, slope, lagoon, patch reef...
 *     'reef_mask'  1 = reef, 0 = not reef
 *   Coverage is the world's shallow tropical reefs — excellent for the
 *   Pacific (verified live for every SPREP country, June 2026).
 *
 * TO LOCALISE: change COUNTRY below.
 **********************************************************************/

// ===== 1. COUNTRY SELECTOR -> REEF AREA ============================
// Larger high islands use the real LSIB boundary; small / atoll nations &
// territories use a point + buffer. The reef area is the coastline grown by
// ~25 km so we capture the fringing and barrier reefs offshore.
var COUNTRY = 'Fiji';   // reef-rich examples: Fiji, Papua New Guinea, Solomon Islands, Palau, Kiribati
var LSIB_NAMES = {
  'Fiji':'Fiji', 'Papua New Guinea':'Papua New Guinea', 'Solomon Islands':'Solomon Is',
  'Vanuatu':'Vanuatu', 'Samoa':'Samoa'};
var POINT_AOI = {
  'Tonga':[-174.8,-20.0,300000], 'Palau':[134.58,7.5,120000], 'Tuvalu':[178.5,-7.8,350000],
  'Kiribati':[173.0,1.4,500000], 'Nauru':[166.93,-0.52,40000], 'Niue':[-169.87,-19.05,40000],
  'Cook Islands':[-159.78,-21.23,300000], 'Marshall Islands':[169.0,8.0,600000],
  'Federated States of Micronesia':[158.21,6.92,500000]};
var land;
if (LSIB_NAMES[COUNTRY]) {
  land = ee.FeatureCollection('USDOS/LSIB_SIMPLE/2017')
           .filter(ee.Filter.eq('country_na', LSIB_NAMES[COUNTRY])).geometry();
} else {
  var pt = POINT_AOI[COUNTRY];
  land = ee.Geometry.Point([pt[0], pt[1]]).buffer(pt[2]);
}
var aoi = land.bounds().buffer(25000);   // reef area = coastline + ~25 km
Map.centerObject(aoi, 9);

// ===== 2. LOAD THE ALLEN CORAL ATLAS ===============================
var aca = ee.Image('ACA/reef_habitat/v2_0').clip(aoi);

// ===== 3. BENTHIC HABITAT MAP (what the seafloor is made of) =======
// The classes are non-sequential codes, so we remap them to 1-6 for a clean
// palette. selfMask() hides everything that is not mapped reef.
var benthicNames  = ['Sand', 'Rubble', 'Rock', 'Seagrass', 'Coral/Algae', 'Microalgal mats'];
var benthicColors = ['ffffbe', 'e0d05e', 'b19c3a', '668438', 'ff6161', '9bcc4f'];
var benthic = aca.select('benthic')
  .remap([11, 12, 13, 14, 15, 18], [1, 2, 3, 4, 5, 6]).selfMask();
Map.addLayer(benthic, {min: 1, max: 6, palette: benthicColors}, 'Benthic habitat');

// ===== 4. GEOMORPHIC ZONES (the shape/structure of the reef) =======
var geoColors = ['77d0fc','2ca2f9','c5a7cb','92739d','614272','fbdefb',
                 '10bda6','288471','cd6812','befbff','ffba15'];
var geomorphic = aca.select('geomorphic')
  .remap([11,12,13,14,15,16,21,22,23,24,25], [1,2,3,4,5,6,7,8,9,10,11]).selfMask();
Map.addLayer(geomorphic, {min: 1, max: 11, palette: geoColors},
             'Geomorphic zones', false);   // off by default — tick it on to view

// ===== 5. LEGEND for the benthic habitat map =======================
var legend = ui.Panel({style: {position: 'bottom-left', padding: '8px', backgroundColor: 'white'}});
legend.add(ui.Label('Benthic habitat', {fontWeight: 'bold', margin: '0 0 4px 0'}));
benthicNames.forEach(function (name, i) {
  legend.add(ui.Panel(
    [ui.Label('', {backgroundColor: '#' + benthicColors[i], padding: '8px', margin: '2px'}),
     ui.Label(name, {margin: '3px 6px'})],
    ui.Panel.Layout.flow('horizontal')));
});
Map.add(legend);

// ===== 6. HOW MUCH CORAL? (habitat area in km²) ====================
var km2 = ee.Image.pixelArea().divide(1e6);
function classAreaKm2(value) {
  return km2.updateMask(aca.select('benthic').eq(value)).reduceRegion({
    reducer: ee.Reducer.sum(), geometry: aoi, scale: 30, maxPixels: 1e12, bestEffort: true
  }).get('area');
}
print('Coral/Algae habitat in ' + COUNTRY + ' (km²):', classAreaKm2(15));
print('Seagrass habitat in ' + COUNTRY + ' (km²):',     classAreaKm2(14));
print('Total mapped reef in ' + COUNTRY + ' (km²):',
  km2.updateMask(aca.select('benthic').gt(0)).reduceRegion({
    reducer: ee.Reducer.sum(), geometry: aoi, scale: 30, maxPixels: 1e12, bestEffort: true}).get('area'));

// ===== 7. EXPORT the benthic habitat map (optional) ================
// 5 m native; export at 10 m to keep the file manageable (raise `scale` if it
// is too large, or shrink the buffer in step 1 to a single island).
Export.image.toDrive({
  image: aca.select('benthic'),
  description: COUNTRY.replace(/[ ,]/g, '_') + '_ACA_Benthic',
  folder: 'GEE_Workshop_2026',
  fileNamePrefix: COUNTRY.replace(/[ ,]/g, '_') + '_aca_benthic',
  scale: 10, region: aoi.bounds(), maxPixels: 1e13
});

File: scripts/python/05_coral_reefs_aca.py
"""
05_coral_reefs_aca.py  —  CORAL REEF HABITATS (Python / geemap)
SPREP / UNEP GEE Climate Workshop 2026

Python equivalent of javascript/05_coral_reefs_aca.js.
Run in Google Colab or Jupyter after 00_setup_geemap.py.

Maps the seafloor habitats (coral, seagrass, sand, ...) and reef zones around
your country from the Allen Coral Atlas, and saves a bar chart of how much of
each habitat your country has.

Dataset: Allen Coral Atlas (ACA/reef_habitat/v2_0)
  ee.Image, 5 m, mapped 2018-2021. Bands:
    'benthic'    seafloor cover (Sand, Rubble, Rock, Seagrass, Coral/Algae,
                 Microalgal Mats)
    'geomorphic' reef zone (reef flat, crest, slope, lagoon, patch reef...)
    'reef_mask'  1 = reef, 0 = not reef
  Covers the world's shallow tropical reefs — verified live for every SPREP
  country (June 2026).
"""

import ee
import geemap
import pandas as pd
import matplotlib.pyplot as plt
# Shared country selector (the 14 SPREP member countries). In Colab / Jupyter
# this helper file may not be present, so fetch it from the workshop repo if so.
try:
    from _pacific_aoi import get_country
except ModuleNotFoundError:
    import urllib.request
    urllib.request.urlretrieve(
        'https://raw.githubusercontent.com/tavitas/sprep-gee-climate-workshop/'
        'main/scripts/python/_pacific_aoi.py', '_pacific_aoi.py')
    from _pacific_aoi import get_country

ee.Initialize(project='your-project-id')   # <-- your registered project

# ===== 1. SETTINGS =====
COUNTRY = 'Fiji'   # reef-rich examples: 'Papua New Guinea', 'Solomon Islands', 'Palau'

land = get_country(COUNTRY)                  # ee.Geometry
aoi = land.bounds().buffer(25000)            # reef area = coastline + ~25 km

# ===== 2. LOAD THE ALLEN CORAL ATLAS =====
aca = ee.Image('ACA/reef_habitat/v2_0').clip(aoi)

# ===== 3. BENTHIC HABITAT MAP =====
# Remap the non-sequential class codes to 1-6 for a clean palette.
benthic_codes  = [11, 12, 13, 14, 15, 18]
benthic_names  = ['Sand', 'Rubble', 'Rock', 'Seagrass', 'Coral/Algae', 'Microalgal mats']
benthic_colors = ['ffffbe', 'e0d05e', 'b19c3a', '668438', 'ff6161', '9bcc4f']
benthic = aca.select('benthic').remap(benthic_codes, [1, 2, 3, 4, 5, 6]).selfMask()

Map = geemap.Map()
Map.centerObject(aoi, 9)
Map.addLayer(benthic, {'min': 1, 'max': 6, 'palette': benthic_colors}, 'Benthic habitat')
Map.add_legend(title='Benthic habitat',
               labels=benthic_names, colors=['#' + c for c in benthic_colors])
Map   # displays in a notebook

# ===== 4. HOW MUCH OF EACH HABITAT? (area in km²) =====
km2 = ee.Image.pixelArea().divide(1e6)


def class_area_km2(code):
    return km2.updateMask(aca.select('benthic').eq(code)).reduceRegion(
        reducer=ee.Reducer.sum(), geometry=aoi, scale=30,
        maxPixels=int(1e12), bestEffort=True).get('area')


areas = ee.Dictionary(
    {name: class_area_km2(code) for name, code in zip(benthic_names, benthic_codes)}
).getInfo()
df = (pd.DataFrame([(n, areas.get(n) or 0) for n in benthic_names], columns=['habitat', 'km2'])
      .sort_values('km2', ascending=True))

plt.figure(figsize=(8, 4))
plt.barh(df['habitat'], df['km2'], color=['#' + benthic_colors[benthic_names.index(h)] for h in df['habitat']])
plt.title(f'Coral reef benthic habitat in {COUNTRY} (Allen Coral Atlas)')
plt.xlabel('Area (km²)')
plt.tight_layout()
plt.savefig(f'{COUNTRY}_coral_habitat.png', dpi=150)
print(f'Saved chart: {COUNTRY}_coral_habitat.png')
for _, r in df.sort_values('km2', ascending=False).iterrows():
    print(f"  {r['habitat']:<16} {r['km2']:,.1f} km²")

# ===== 5. EXPORT the benthic habitat map (optional) =====
# 5 m native; export at 10 m to keep the file manageable.
task = ee.batch.Export.image.toDrive(
    image=aca.select('benthic'),
    description=f'{COUNTRY}_ACA_Benthic',
    folder='GEE_Workshop_2026',
    fileNamePrefix=f'{COUNTRY.replace(" ", "_")}_aca_benthic',
    scale=10,
    region=aoi.bounds(),
    maxPixels=int(1e13))
task.start()
print('Export started — check the Tasks panel at https://code.earthengine.google.com')
